src/main/kotlin/eu/europa/ec/eudi/pidissuer/adapter/out/jose/ResolveDidUrl.kt

```diff
@@ -0,0 +1,155 @@
+                separator = ", ",
+                transform = { it.typeName },
+            )
+        }'."
+    }
+
+    val expectedKeySizes = mapOf(
+        Multicodec.SECP256K1_PUB to 33,
+        Multicodec.X25519_PUB to 32,
+        Multicodec.ED25519_PUB to 32,
+        Multicodec.P256_PUB to 33,
+        Multicodec.P384_PUB to 49,
+    )
+    expectedKeySizes[type]?.let { expectedKeySize ->
+        require(expectedKeySize == key.size) {
+            "Expected a key size of '$expectedKeySize' for type '${type.typeName}'. Got '${key.size}' instead."
+        }
+    }
+
+    fun decodeOctetPublicKey(curve: Curve, publicKey: ByteArray): OctetKeyPair =
+        OctetKeyPair.Builder(curve, Base64URL.encode(publicKey)).build()
+
+    fun decodeEcPublicKey(curve: Curve, publicKey: ByteArray): ECKey {
+        val spec = ECNamedCurveTable.getParameterSpec(curve.stdName)
+        val point = spec.curve.decodePoint(publicKey)
+        val keySpec = ECPublicKeySpec(point, spec)
+        val factory = KeyFactory.getInstance("EC", BouncyCastleProviderSingleton.getInstance())
+        val ecPublicKey = factory.generatePublic(keySpec) as ECPublicKey
+        return ECKey.Builder(curve, ecPublicKey).build()
+    }
+
+    fun decodeRsaPublicKey(encodedPublicKey: ByteArray): RSAKey {
+        val ans1PublicKey = ANS1RSAPublicKey.getInstance(encodedPublicKey)
+        val keySpec = RSAPublicKeySpec(ans1PublicKey.modulus, ans1PublicKey.publicExponent)
+        val factory = KeyFactory.getInstance("RSA", BouncyCastleProviderSingleton.getInstance())
+        val publicKey = factory.generatePublic(keySpec) as RSAPublicKey
+        return RSAKey.Builder(publicKey).build()
+    }
+
+    return when {
+        Multicodec.SECP256K1_PUB == type -> decodeEcPublicKey(Curve.SECP256K1, key)
+        Multicodec.X25519_PUB == type -> decodeOctetPublicKey(Curve.X25519, key)
+        Multicodec.ED25519_PUB == type -> decodeOctetPublicKey(Curve.Ed25519, key)
+        Multicodec.P256_PUB == type -> decodeEcPublicKey(Curve.P_256, key)
+        Multicodec.P384_PUB == type -> decodeEcPublicKey(Curve.P_384, key)
+        Multicodec.P521_PUB == type -> decodeEcPublicKey(Curve.P_521, key)
+        Multicodec.RSA_PUB == type -> decodeRsaPublicKey(key)
+        else -> throw IllegalArgumentException(
+            "Unsupported type '${type.typeName}'. Expected on of '${
+                supportedTypes.joinToString(
+                    separator = ", ",
+                    transform = { it.typeName },
+                )
+            }'.",
+        )
+    }
+}
+
+private fun resolveDidJwk(url: DIDURL): JWK {
+    require(url.did.methodName == "jwk") {
+        "Expected 'key' method. Got '${url.did.methodName}' instead."
+    }
+    require(url.fragment.isNullOrBlank()) {
+        "Invalid fragment. Expected no fragment but got '${url.fragment}' instead."
+    }
+
+    return JWK.parse(Base64URL.from(url.did.methodSpecificId).decodeToString())
```

Comment:
I think that there should be also the requirement that the JWK should not represent a private key